### PR TITLE
Export Telemetry data to different exporters

### PR DESCRIPTION
**Pull request by lkomali**
This PR modifies the existing output writer and exporter classes to handle telemetry data.

Below is how the output looks like

1. Console output with verbose
image

2. CSV output
image

3. JSON output
image

### CODE REVIEW

File: genai-perf/genai_perf/export_data/csv_exporter.py [hunk 1]
```diff
@@ -73,6 +96,11 @@ def export(self) -> None:
             self._write_request_metrics(writer)
             writer.writerow([])
             self._write_system_metrics(writer)
+            if self._telemetry_stats is not None:
```
Comment (dyastremsky) [inline genai-perf/genai_perf/export_data/csv_exporter.py:99]: You may want to put this into its own function. We want to keep the telemetry code as contained as possible so that it doesn't start sprawling throughout the code base and unrelated code.

File: genai-perf/genai_perf/export_data/csv_exporter.py [hunk 2]
```diff
@@ -100,6 +128,55 @@ def _write_system_metrics(self, csv_writer) -> None:
             value = self._stats[metric.name]["avg"]
             csv_writer.writerow([metric_str, f"{value:.2f}"])
 
+    def _write_telemetry_aggregated_metrics(self, csv_writer) -> None:
```
Comment (debermudez) [inline genai-perf/genai_perf/export_data/csv_exporter.py:131]: The breaks in the csv we already had in place cause a lot of problems for downstream consumers.
Ideally, they would use the json, but I would update this to use the existing tables as much as possible.

File: genai-perf/tests/test_console_exporter.py [hunk 3]
```diff
@@ -214,19 +223,22 @@ def test_valid_goodput(self, monkeypatch, capsys) -> None:
...
+            "┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┳━━━━━━━┳━━━━━━━┳━━━━━━━┳━━━━━━━┳━━━━━━━┓\n"
+            "┃ Metric                      ┃    avg ┃   min ┃   max ┃   p99 ┃   p90 ┃   p75 ┃\n"
+            "┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━╇━━━━━━━╇━━━━━━━╇━━━━━━━╇━━━━━━━╇━━━━━━━┩\n"
+            "│ Time to first token (ms)    │   8.00 │  7.00 │  9.00 │  8.98 │  8.80 │  8.50 │\n"
+            "│ Inter token latency (ms)    │  11.00 │ 10.00 │ 12.00 │ 11.98 │ 11.80 │ 11.50 │\n"
+            "│ Request latency (ms)        │   5.00 │  4.00 │  6.00 │  5.98 │  5.80 │  5.50 │\n"
+            "│ Output sequence length      │   2.00 │  1.00 │  3.00 │  2.98 │  2.80 │  2.50 │\n"
+            "│ Input sequence length       │   6.00 │  5.00 │  7.00 │  6.98 │  6.80 │  6.50 │\n"
+            "│ Output token throughput     │ 456.00 │   N/A │   N/A │   N/A │   N/A │   N/A │\n"
+            "│ (per sec)                   │        │       │       │       │       │       │\n"
```
Comment (debermudez) [inline genai-perf/tests/test_console_exporter.py:235]: Why is the text wrapping now?

File: genai-perf/tests/test_console_exporter.py [hunk 4]
```diff
@@ -165,9 +174,9 @@ def test_embedding_output(self, monkeypatch, capsys) -> None:
         expected_content = (
             "                   NVIDIA GenAI-Perf | Embeddings Metrics                   \n"
             "┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┳━━━━━━┳━━━━━━┳━━━━━━┳━━━━━━┳━━━━━━┓\n"
-            "┃                    Statistic ┃    avg ┃  min ┃  max ┃  p99 ┃  p90 ┃  p75 ┃\n"
+            "┃ Metric                       ┃    avg ┃  min ┃  max ┃  p99 ┃  p90 ┃  p75 ┃\n"
```
Comment (debermudez) [inline genai-perf/tests/test_console_exporter.py:177]: @harryskim need your input here.
Is this title better as metric or statistic?

File: genai-perf/tests/test_console_exporter.py [hunk 5]
```diff
@@ -214,19 +223,22 @@ def test_valid_goodput(self, monkeypatch, capsys) -> None:
...
-            "│            Output sequence length │ 2.00 │ 1.00 │ 3.00 │ 2.98 │  2.80 │ 2.50 │\n"
-            "│             Input sequence length │ 6.00 │ 5.00 │ 7.00 │ 6.98 │  6.80 │ 6.50 │\n"
-            "│ Output token throughput (per sec) │ 456… │  N/A │  N/A │  N/A │   N/A │  N/A │\n"
-            "│      Request throughput (per sec) │ 123… │  N/A │  N/A │  N/A │   N/A │  N/A │\n"
-            "│         Request goodput (per sec) │ 100… │  N/A │  N/A │  N/A │   N/A │  N/A │\n"
-            "└───────────────────────────────────┴──────┴──────┴──────┴──────┴───────┴──────┘\n"
+            "┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┳━━━━━━━┳━━━━━━━┳━━━━━━━┳━━━━━━━┳━━━━━━━┓\n"
+            "┃ Metric                      ┃    avg ┃   min ┃   max ┃   p99 ┃   p90 ┃   p75 ┃\n"
+            "┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━╇━━━━━━━╇━━━━━━━╇━━━━━━━╇━━━━━━━╇━━━━━━━┩\n"
+            "│ Time to first token (ms)    │   8.00 │  7.00 │  9.00 │  8.98 │  8.80 │  8.50 │\n"
```
Comment (debermudez) [inline genai-perf/tests/test_console_exporter.py:229]: @harryskim thoughts on the left alignment for the left most column and the right alignment for the others?

File: genai-perf/tests/test_console_exporter.py (hunk 4 above)
Comment (debermudez) [reply genai-perf/tests/test_console_exporter.py:177]: It matches with the title more but it is a change i think we need to run by you.

File: genai-perf/tests/test_json_exporter.py [hunk 6]
```diff
@@ -229,137 +230,139 @@ def test_generate_json_custom_export(
...
-            "ignore_eos": true
-          }
+        {
+            "llm_metrics": {
+                "request_throughput": {
+                "unit": "requests/sec",
+                "avg": "7"
+                },
+                "request_latency": {
+                "unit": "ms",
```
Comment (debermudez) [inline genai-perf/tests/test_json_exporter.py:240]: I would not expect the formatting to change this much.
Are you running the same pre commit hook?

File: genai-perf/tests/test_json_exporter.py [hunk 7]
```diff
@@ -675,10 +679,277 @@ def test_invalid_goodput_json_output(
...
+        config.extra_inputs = parser.get_extra_inputs_as_dict(args)
+        config.artifact_dir = args.artifact_dir
+        json_exporter = JsonExporter(config)
+        assert json_exporter._stats_and_args["telemetry_metrics"] == json.loads(
+            expected_telemetry_json_output
+        )
+
+        json_exporter.export()
+        expected_filename = "profile_export_genai_perf.json"
+        written_data = [
```
Comment (debermudez) [inline genai-perf/tests/test_json_exporter.py:941]: will this ever be more than 1 value?
a list made me think it would be >1 entry but then you assert on 1 entry below.

File: genai-perf/genai_perf/parser.py [hunk 8]
```diff
@@ -882,13 +882,23 @@ def profile_handler(args, extra_args):
...
+        )
+        telemetry_data_collector = None
 
     Profiler.run(
         args=args,
         extra_args=extra_args,
         telemetry_data_collector=telemetry_data_collector,
     )
 
+    print(telemetry_data_collector)
```
[the remark this answers is not in the document]
Comment (lkomali) [reply genai-perf/genai_perf/parser.py:898]: This was for my debug. This print statement should be removed.

**Comment by lkomali** [comment]
This work was tackled in 5 separate PR's. Closing it as it's not required.
